File: go/examples/date-diff-days_test.go
// カード date-diff-days の Contract を検証するテスト
package examples

import (
	"math"
	"testing"
	"time"
)

func diffDays(later, earlier time.Time) int {
	l := time.Date(later.Year(), later.Month(), later.Day(), 0, 0, 0, 0, time.UTC)
	e := time.Date(earlier.Year(), earlier.Month(), earlier.Day(), 0, 0, 0, 0, time.UTC)
	return int(l.Sub(e).Hours() / 24)
}

func TestDateDiffDays(t *testing.T) {
	tokyo, err := time.LoadLocation("Asia/Tokyo")
	if err != nil {
		t.Fatal(err)
	}
	ny, err := time.LoadLocation("America/New_York")
	if err != nil {
		t.Fatal(err)
	}
	later := time.Date(2024, 3, 1, 0, 1, 0, 0, tokyo)
	earlier := time.Date(2024, 2, 29, 23, 59, 0, 0, tokyo)

	t.Run("後の日付、前の日付の順で暦日差。逆順なら負、同じ日なら 0", func(t *testing.T) {
		if got := diffDays(later, earlier); got != 1 {
			t.Errorf("later - earlier = %d", got)
		}
		if got := diffDays(earlier, later); got != -1 {
			t.Errorf("earlier - later = %d", got)
		}
		if got := diffDays(later, later); got != 0 {
			t.Errorf("同じ日 = %d", got)
		}
		if got := diffDays(time.Date(2025, 3, 1, 0, 0, 0, 0, tokyo), time.Date(2024, 3, 1, 0, 0, 0, 0, tokyo)); got != 365 {
			t.Errorf("1 年 = %d", got)
		}
	})

	t.Run("時刻は無視される。実時間の差は 2 分", func(t *testing.T) {
		if got := later.Sub(earlier); got != 2*time.Minute {
			t.Errorf("Sub = %v", got)
		}
		if got := int(later.Sub(earlier).Hours() / 24); got != 0 {
			t.Errorf("実時間ベース = %d", got)
		}
	})

	t.Run("DST 切替日（実時間 23 時間）でも暦日差は 1", func(t *testing.T) {
		d0 := time.Date(2024, 3, 10, 0, 0, 0, 0, ny)
		d1 := time.Date(2024, 3, 11, 0, 0, 0, 0, ny)
		if got := d1.Sub(d0); got != 23*time.Hour {
			t.Errorf("Sub = %v", got)
		}
		if got := diffDays(d1, d0); got != 1 {
			t.Errorf("diffDays = %d", got)
		}
		if got := int(d1.Sub(d0).Hours() / 24); got != 0 {
			t.Errorf("切り捨て = %d", got)
		}
	})

	t.Run("Sub は約 292 年で飽和する", func(t *testing.T) {
		base := time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)
		if got := time.Date(2292, 1, 1, 0, 0, 0, 0, time.UTC).Sub(base); got == time.Duration(math.MaxInt64) {
			t.Errorf("292 年は飽和しないはず")
		}
		if got := time.Date(2293, 1, 1, 0, 0, 0, 0, time.UTC).Sub(base); got != time.Duration(math.MaxInt64) {
			t.Errorf("293 年 = %v", got)
		}
	})

	t.Run("Location が違うとそれぞれのローカル日付で比べる", func(t *testing.T) {
		jst := time.Date(2024, 3, 1, 0, 30, 0, 0, tokyo)
		utc := time.Date(2024, 2, 29, 20, 0, 0, 0, time.UTC)
		if got := diffDays(jst, utc); got != 1 {
			t.Errorf("混在 = %d", got)
		}
		if got := diffDays(jst.In(time.UTC), utc); got != 0 {
			t.Errorf("UTC に揃えた = %d", got)
		}
	})

	t.Run("UTC+14 と UTC-12 で同じローカル日付なら、実時間が離れていても 0", func(t *testing.T) {
		plus14 := time.Date(2024, 3, 1, 23, 0, 0, 0, time.FixedZone("UTC+14", 14*3600))
		minus12 := time.Date(2024, 3, 1, 0, 0, 0, 0, time.FixedZone("UTC-12", -12*3600))
		if got := diffDays(plus14, minus12); got != 0 {
			t.Errorf("+14 - (-12) = %d", got)
		}
		if got := diffDays(minus12, plus14); got != 0 {
			t.Errorf("-12 - (+14) = %d", got)
		}
	})

	t.Run("入力を変更せず、ゼロ値も扱える", func(t *testing.T) {
		before := later
		diffDays(later, earlier)
		if !later.Equal(before) {
			t.Errorf("入力が変更された")
		}
		if got := diffDays(time.Date(1, 1, 2, 0, 0, 0, 0, time.UTC), time.Time{}); got != 1 {
			t.Errorf("zero = %d", got)
		}
	})

	t.Run("日付をスキップするゾーン（Pacific/Apia 2011-12-30）でも暦日差は日付の差", func(t *testing.T) {
		apia, err := time.LoadLocation("Pacific/Apia")
		if err != nil {
			t.Skip("tzdata なし")
		}
		a := time.Date(2011, 12, 29, 12, 0, 0, 0, apia)
		b := time.Date(2011, 12, 31, 12, 0, 0, 0, apia)
		if got := diffDays(b, a); got != 2 {
			t.Errorf("diffDays = %d, want 2", got)
		}
	})

}

File: python/examples/date-diff-days_test.py
"""date-diff-days: date 同士の減算による暦日差の Contract を検証する。"""

from datetime import date, datetime, timedelta, timezone
from zoneinfo import ZoneInfo

import pytest

JST = ZoneInfo("Asia/Tokyo")
NY = ZoneInfo("America/New_York")

LATER = datetime(2024, 3, 1, 0, 1)
EARLIER = datetime(2024, 2, 29, 23, 59)


def test_date_subtraction_returns_timedelta_with_calendar_days() -> None:
    """date - date は timedelta を返し、.days が暦日差。"""
    diff = date(2024, 3, 1) - date(2024, 2, 29)
    assert isinstance(diff, timedelta)
    assert diff == timedelta(days=1)
    assert diff.days == 1
    assert (date(2025, 3, 1) - date(2024, 3, 1)).days == 365


def test_sign_depends_on_operand_order() -> None:
    """左辺が後なら正、前なら負、同じ日なら 0。"""
    assert (LATER.date() - EARLIER.date()).days == 1
    assert (EARLIER.date() - LATER.date()).days == -1
    assert (date(2024, 2, 29) - date(2024, 2, 29)).days == 0


def test_time_of_day_is_ignored() -> None:
    """時刻を捨てるので、2 分差でも日付が変われば 1。同じ日なら時刻に関係なく 0。"""
    assert (LATER.date() - EARLIER.date()).days == 1
    assert (datetime(2024, 3, 1, 23, 59).date() - datetime(2024, 3, 1, 0, 0).date()).days == 0


def test_inputs_are_not_mutated() -> None:
    """入力は変わらない。"""
    a, b = date(2024, 3, 1), date(2024, 2, 29)
    a - b
    assert a == date(2024, 3, 1)
    assert b == date(2024, 2, 29)


def test_datetime_difference_days_floors_toward_negative_infinity() -> None:
    """datetime - datetime の .days は 24 時間単位で −∞ 方向に丸める。"""
    assert (LATER - EARLIER) == timedelta(minutes=2)
    assert (LATER - EARLIER).days == 0
    assert (EARLIER - LATER).days == -1
    assert (datetime(2024, 3, 1, 12) - datetime(2024, 3, 1, 0)).days == 0
    assert (datetime(2024, 3, 1, 0) - datetime(2024, 3, 1, 12)).days == -1


def test_floor_division_by_one_day_also_floors() -> None:
    """Alternatives の timedelta // timedelta(days=1) も −∞ 方向の丸め。"""
    assert (LATER - EARLIER) // timedelta(days=1) == 0
    assert (EARLIER - LATER) // timedelta(days=1) == -1
    assert (LATER - EARLIER).total_seconds() / 86400 == pytest.approx(2 / 1440)


def test_mixing_aware_and_naive_datetime_raises() -> None:
    """aware と naive の datetime を直接引くと TypeError。.date() にすれば引ける。"""
    aware = datetime(2024, 3, 1, tzinfo=JST)
    naive = datetime(2024, 3, 1)
    with pytest.raises(TypeError):
        aware - naive
    assert (aware.date() - naive.date()).days == 0


def test_date_and_datetime_cannot_be_subtracted() -> None:
    """date - datetime、datetime - date は TypeError。"""
    with pytest.raises(TypeError):
        date(2024, 3, 1) - datetime(2024, 2, 29)  # type: ignore[operator]
    with pytest.raises(TypeError):
        datetime(2024, 3, 1) - date(2024, 2, 29)  # type: ignore[operator]


def test_adding_timedelta_is_the_inverse() -> None:
    """date + timedelta で逆演算になる。"""
    assert EARLIER.date() + timedelta(days=1) == LATER.date()
    assert LATER.date() - timedelta(days=1) == EARLIER.date()


def test_aware_compare_by_instant_but_dates_are_local() -> None:
    """aware 同士の比較（<）は実時刻で行われる。.date() にすると各自のローカル日付になるので、揃えるなら astimezone してから。"""
    jst = datetime(2024, 3, 1, 0, 30, tzinfo=JST)
    utc = datetime(2024, 2, 29, 20, 0, tzinfo=timezone.utc)
    assert jst < utc
    assert (jst.date() - utc.date()).days == 1
    assert (jst.astimezone(timezone.utc).date() - utc.date()).days == 0


def test_dst_day_counts_as_one_calendar_day() -> None:
    """DST 切替日（23 時間）も暦日では 1 日。"""
    a = datetime(2024, 3, 11, 0, 0, tzinfo=NY)
    b = datetime(2024, 3, 9, 0, 0, tzinfo=NY)
    assert (a.date() - b.date()).days == 2
    assert a.astimezone(timezone.utc) - b.astimezone(timezone.utc) == timedelta(days=1, hours=23)
